### ISSUE TITLE
Impossible to fetch artifact for the build that is no longer in queue

### ISSUE THREAD

**Issue by wtrocki** [jenkinsci/java-client-api#213]
Queue items are removed after build is finished. Current api allows only to fetch artifact for the queue item which makes this solution extremely inflexible. 

To fetch artifact any time after build new api method should be introduced that would allow to specify build number and job name as input.

**Comment by khmarbaise** [comment]
Maybe I misunderstand your request but what about:

```java
JenkinsServer js = new JenkinsServer(URI.create("http://jenkins/"));
JobWithDetails job = js.getJob("test-results");
Build buildByNumber = job.getBuildByNumber(1);
```
or what about:
```java
JenkinsServer js = new JenkinsServer(URI.create("http://localhost:27100/buildserver/"));
JobWithDetails job = js.getJob("test-results");

List<Build> builds = job.getBuilds();
List<Build> allBuilds = job.getAllBuilds();
List<Build> buildRanges = job.getAllBuilds(Range.build().from(5).to(10));
```

**Comment by wtrocki** [comment]
Yes. That's the workaround I'm using at the moment, but in the background it would make two requests. Feel free to close this issue.

**Comment by khmarbaise** [comment]
If you take a look into the code you will see that only `getJob(..)` will make a REST API call whereas `getBuildByNumber(..)` will not do so.

If you use `getAllBuilds()...` or `getAllBuilds(Range...)` than a supplemental call will be done...

**Comment by wtrocki** [comment]
Yes, but that object is not useful. In real word scenario you always fetching BuildWithDetails to get information about artifacts etc. Fix in PR is really simple and it would make API much cleaner. It may break existing object oriented concept.

**Comment by khmarbaise** [comment]
It sounds like I misunderstand what you have written...

```java
Build buildByNumber = job.getBuildByNumber(1);
```
If you have `Build` you can fetch the needed information via
```java
BuildWithDetails details = buildByNumber.details()
```
BTW: Can you give a code example of what you like to do or what you are doing ? So I get a better impression of what we are talking about?

**Comment by wtrocki** [comment]
Sorry for confusion. I think we are ok with this one additional call.
That's our current implementation:
https://github.com/aerogear/digger-java/blob/master/src/main/java/org/aerogear/digger/services/ArtifactsService.java#L73-L79

After merging this PR https://github.com/jenkinsci/java-client-api/pull/214 we can just:
{code}
BuildWithDetails build = client.getBuild("test",1);
{code}

I think that aproach not aligned with the way this client is build and we can close this issue.
Sorry for confusion.

**Comment by khmarbaise** [comment]
You can of course use something like this: `js.getJob("non-maven-test").getBuildByNumber(1).details();`? Or do you think it's more convenient to have supplemental methods with parameters you have suggested?
